8254129: IR Test Framework to support regex-based matching on the IR in JTreg compiler tests

chhagedorn opened the pull request:
This RFE provides an IR test framework to perform regex-based checks on the C2 IR shape of test methods emitted by the VM flags `-XX:+PrintIdeal` and `-XX:+PrintOptoAssembly`. The framework can also be used for other non-IR matching (and non-compiler) tests by providing easy to use annotations for commonly used testing patterns and compiler control flags.

The framework is based on the ideas of the currently present IR test framework in [Valhalla](https://github.com/openjdk/valhalla/blob/e9c78ce4fcfd01361c35883e0d68f9ae5a80d079/test/hotspot/jtreg/compiler/valhalla/inlinetypes/InlineTypeTest.java) (mainly implemented by @TobiHartmann) which is being used with great success. This new framework aims to replace the old one in Valhalla at some point.

A detailed description about how this new IR test framework works and how it is used is provided in the [README.md](https://github.com/chhagedorn/jdk/blob/aa005f384a4567c6c0b5f08f7c5df57f705dc540/test/lib/jdk/test/lib/hotspot/ir_framework/README.md) file and in the [Javadocs](https://github.com/chhagedorn/jdk/blob/aa005f384a4567c6c0b5f08f7c5df57f705dc540/test/lib/jdk/test/lib/hotspot/ir_framework/doc/jdk/test/lib/hotspot/ir_framework/package-summary.html) written for the framework classes.

To finish a first version of this framework for JDK 17, I decided to leave some improvement possibilities and ideas to be followed up on in additional RFEs. Some ideas are mentioned in "Future Work" in [README.md](https://github.com/chhagedorn/jdk/blob/aa005f384a4567c6c0b5f08f7c5df57f705dc540/test/lib/jdk/test/lib/hotspot/ir_framework/README.md) and were also created as subtasks of this RFE.

Testing (also described in "Internal Framework Tests in [README.md](https://github.com/chhagedorn/jdk/blob/aa005f384a4567c6c0b5f08f7c5df57f705dc540/test/lib/jdk/test/lib/hotspot/ir_framework/README.md)):
There are various tests to verify the correctness of the test framework which can be found as JTreg tests in the [tests](https://github.com/chhagedorn/jdk/tree/aa005f384a4567c6c0b5f08f7c5df57f705dc540/test/lib/jdk/test/lib/hotspot/ir_framework/tests) folder. Additional testing was performed by converting all compiler Inline Types test of project Valhalla (done by @katyapav in [JDK-8263024](https://bugs.openjdk.java.net/browse/JDK-8263024)) that used the old framework to the new framework. This provided additional testing for the framework itself. We ran the converted tests with all the flag settings used in hs-tier1-9 and hs-precheckin-comp. For sanity checking, this was also done with a sample IR test in mainline.

Some stats about the framework code added to [ir_framework](https://github.com/chhagedorn/jdk/tree/aa005f384a4567c6c0b5f08f7c5df57f705dc540/test/lib/jdk/test/lib/hotspot/ir_framework):

-  without the [Javadocs files](https://github.com/chhagedorn/jdk/tree/aa005f384a4567c6c0b5f08f7c5df57f705dc540/test/lib/jdk/test/lib/hotspot/ir_framework/doc) : 60 changed files, 13212 insertions, 0 deletions.
   - without the [tests](https://github.com/chhagedorn/jdk/tree/aa005f384a4567c6c0b5f08f7c5df57f705dc540/test/lib/jdk/test/lib/hotspot/ir_framework/tests)  and [examples](https://github.com/chhagedorn/jdk/tree/aa005f384a4567c6c0b5f08f7c5df57f705dc540/test/lib/jdk/test/lib/hotspot/ir_framework/examples) folder: 40 files changed, 6781 insertions
      - comments: 2399 insertions (calculated with `git diff --cached !(tests|examples) | grep -c -E "(^[+-]\s*(\/)?\*)|(^[+-]\s*\/\/)"`)
      - which leaves 4382 lines of code inserted 

Big thanks to:
- @TobiHartmann for all his help by discussing the new framework and for providing insights from his IR test framework in Valhalla.
- @katyapav for converting the Valhalla tests to use the new framework which found some harder to catch bugs in the framework and also some actual C2 bugs.
- @iignatev for helping to simplify the framework usage with JTreg and with the framework internal VM calling structure.
- and others who provided valuable feedback.

Thanks,
Christian

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8254129](https://bugs.openjdk.java.net/browse/JDK-8254129): IR Test Framework to support regex-based matching on the IR in JTreg compiler tests


### Reviewers
 * [Igor Ignatyev](https://openjdk.java.net/census#iignatyev) (@iignatev - **Reviewer**) ⚠️ Review applies to df7576f7d68a4cba0431e96f336bf57ae025d70b


### Contributors
 * Christian Hagedorn `<chagedorn@openjdk.org>`
 * Tobias Hartmann `<thartmann@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3508/head:pull/3508` \
`$ git checkout pull/3508`

Update a local copy of the PR: \
`$ git checkout pull/3508` \
`$ git pull https://git.openjdk.java.net/jdk pull/3508/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3508`

View PR using the GUI difftool: \
`$ git pr show -t 3508`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3508.diff">https://git.openjdk.java.net/jdk/pull/3508.diff</a>

</details>

chhagedorn commented:
/label remove javadoc

chhagedorn commented:
/label add hotspot-compiler
/label add hotspot

chhagedorn commented:
/contributor add @TobiHartmann

iignatev commented:
Hi Christian,

kudos to you, Tobias, Katya, and all involved, I have really high hopes that this framework will improve the quality of both JVM and our testing.  

I'll look at the code later, a few general comments: 
 - although having javadoc for testlibraries is highly desirable, I don't think we should check in the generated HTML files
 - the same goes for `README.html` generated from `README.md`
 - this library is hotspot-centric, I highly doubt that it will be used by any tests outside of the hotspot test base, hence the more appropriate location for it would be inside `test/hotspot/jtreg/testlibrary`.
 - I assume `test/lib/jdk/test/lib/hotspot/ir_framework/tests/` are the tests for the framework, in that case they should be in `test/lib-test`, if we stick to `test/lib` as the location for the library, or in `test/hotspot/jtreg/testlibrary_tests`, if we move it to `test/hotspot`


Thanks,
-- Igor

vnkozlov reviewed (commented):
I understand the desire to use this framework for more then IR verification but it is main goal.
Why you did not place it into `test/lib/sun/hotspot/code/`?

I also don't think you should include javadoc generated files into these changes.

dougxc commented on test/lib/jdk/test/lib/hotspot/ir_framework/README.html:
```diff
@@ -0,0 +1,160 @@
+<h1>IR Test Framework</h1>
```
Just curious: why HTML instead of markdown?

iignatev commented:
> Why you did not place it into `test/lib/sun/hotspot/code/`?
there is an RFE to rename `sun.hotspot` testlibrary classes, so I wouldn't put new stuff there.

iignatev commented on test/lib/jdk/test/lib/hotspot/ir_framework/TestFramework.java:
```diff
@@ -0,0 +1,1226 @@
+        run(walker.getCallerClass());
+    }
+
+    /**
```
I'm not sure how much of the benefits all these different `run.*` bring. I fully understand the desire to simplify the most common use-case (i.e. no-arg `run`), but for the rest I'd assume the users will be capable of writing, for example, `new TestFramework(testClass).addScenarios(scenarios).start();` instead of `TestFramework .runWithScenarios(testClass, scenarios);`

iignatev replied:
> I suggest to keep run() and maybe also runWithFlags() as (probably?) most common use-case.

I guess you are right, `run()` and `runWithFlags(String[])` will be most-commonly used ones.

iignatev commented on test/lib/jdk/test/lib/hotspot/ir_framework/TestFramework.java:
```diff
@@ -0,0 +1,1226 @@
+    private static String lastTestVMOutput;
+
+    private final Class<?> testClass;
+    private List<Class<?>> helperClasses = null;
+    private List<Scenario> scenarios = null;
+    private final List<String> flags = new ArrayList<>();
```
why did you decide to eagerly create list for `flags`, but not for `scenarios` / `helpersClasses`?

iignatev commented on test/lib/jdk/test/lib/hotspot/ir_framework/TestFramework.java:
```diff
@@ -0,0 +1,1226 @@
+        }
+
+        for (Class<?> helperClass : helperClasses) {
+            TestRun.check(!this.helperClasses.contains(helperClass), "Cannot add the same class twice: " + helperClass);
```
won't it be easy to use `Set` to store `helperClasses`?

iignatev replied:
you might also want to update javadoc for this method to state that there can be duplicates.

iignatev replied:
you don't have to ignore duplicates, you can still throw an exception if you encounter one. I, personally, don't think we should force users to avoid dups (as I don't yet? see any issues w/ duplicated helpers), but if you think we need to enforce that *and* make it users' responsibilities then you should continue to throw an exception. in both cases, javadoc should be updated to reflect that duplicates are ignored or that an exception is raised in case a duplicate was encountered.

iignatev commented on test/lib/jdk/test/lib/hotspot/ir_framework/TestFramework.java:
```diff
@@ -0,0 +1,1226 @@
+            this.helperClasses = new ArrayList<>();
+        }
+
+        for (Class<?> helperClass : helperClasses) {
```
nit: I'd use `var` here

iignatev commented on test/lib/jdk/test/lib/hotspot/ir_framework/TestFramework.java:
```diff
@@ -0,0 +1,1226 @@
+            try {
+                start(null);
+            } catch (TestVMException e) {
+                System.err.println("\n" + e.getExceptionInfo());
```
you shouldn't use "\n", as it might not be the right line-separator. you can either do:
```suggestion
                System.err.println();
                System.err.println(e.getExceptionInfo());
```
or
```suggestion
                System.err.printf("%n%s%n", e.getExceptionInfo());
```

iignatev replied:
sure you case use `System.lineSeparator()`, it's just a matter of personal choice/style

chhagedorn replied:
:+1: I fixed them.

iignatev commented on test/lib/jdk/test/lib/hotspot/ir_framework/TestFramework.java:
```diff
@@ -0,0 +1,1226 @@
+     */
+    private void maybeDisableIRVerificationCompletely() {
+        if (VERIFY_IR) {
+            VERIFY_IR = hasIRAnnotations();
+            if (!VERIFY_IR) {
+                System.out.println("IR verification disabled due to test " + testClass + " not specifying any @IR annotations");
+                return;
+            }
+
+            VERIFY_IR = Platform.isDebugBuild() && !Platform.isInt() && !Platform.isComp();
+            if (!VERIFY_IR) {
+                System.out.println("IR verification disabled due to not running a debug build (required for PrintIdeal" +
+                                   "and PrintOptoAssembly), running with -Xint, or -Xcomp (use warm-up of 0 instead)");
+                return;
+            }
```
I'd reorder them as "platform" checks are faster than `hasIRAnnotations` check and, I'd guess, will be a more common culprit to disable IR verification.

iignatev commented on test/lib/jdk/test/lib/hotspot/ir_framework/TestFramework.java:
```diff
@@ -0,0 +1,1226 @@
+    /**
+     * Disable IR verification completely in certain cases.
+     */
+    private void maybeDisableIRVerificationCompletely() {
```
nit: 
```suggestion
    private void disableIRVerificationIfNotFeasible() {
```

iignatev commented on test/lib/jdk/test/lib/hotspot/ir_framework/TestFramework.java:
```diff
@@ -0,0 +1,1226 @@
+            if (scenario != null) {
+                errorMsg = getScenarioTitleAndFlags(scenario);
+            }
+            if (e instanceof IRViolationException) {
+                IRViolationException irException = (IRViolationException) e;
```
nit:
```suggestion
            if (e instanceof IRViolationException irException) {
```

iignatev commented on test/lib/jdk/test/lib/hotspot/ir_framework/TestFramework.java:
```diff
@@ -0,0 +1,1226 @@
+                System.out.println((scenario != null ? "Scenario #" + scenario.getIndex() + " - " : "")
+                                   + "Compilation(s) of failed matche(s):");
+                System.out.println(irException.getCompilations());
+                builder.append(errorMsg).append("\n").append(irException.getExceptionInfo()).append(e.getMessage());
```
as `builder.toString` will be printed out to cout/cerr, you should use platform-specific line-separator instead of `\n`

iignatev reviewed (changes requested):
Hi Christian,

here is the 1st portion of my comments. I'll return to reviewing it 1st thing tomorrow...

Cheers,
-- Igor

iignatev commented on test/lib/jdk/test/lib/hotspot/ir_framework/IRMatcher.java:
```diff
@@ -0,0 +1,562 @@
+     */
+    private Map<String, Integer[]>  parseIREncoding(String irEncoding) {
+        Map<String, Integer[]> irRulesMap = new HashMap<>();
+        String patternString = "(?<=" + IREncodingPrinter.START + "\\R)[\\s\\S]*(?=" + IREncodingPrinter.END + ")";
+        Pattern pattern = Pattern.compile(patternString);
```
`patternString` is effectively a constant here, but you will compile it into `j.u.Pattern` on each invocation of `parseIREncoding`, it might be a better idea to introduce a static field that holds a precompiled patter.

iignatev commented on test/lib/jdk/test/lib/hotspot/ir_framework/IRMatcher.java:
```diff
@@ -0,0 +1,562 @@
+     * and IR encoding from the test VM specifies that this method has @IR rules to be checked).
+     */
+    private void addTestMethodCompileId(Map<Integer, String> compileIdMap, String line) {
+        Pattern pattern = Pattern.compile("compile_id='(\\d+)'.*" + Pattern.quote(testClass.getCanonicalName()) + " (\\S+)");
+        Matcher matcher = pattern.matcher(line);
```
similarly to `parseIREncoding`, `pattern` here can be precomputed in `IRMatcher::IRMatcher` and stored into a final instance field.

iignatev commented on test/lib/jdk/test/lib/hotspot/ir_framework/IRMatcher.java:
```diff
@@ -0,0 +1,562 @@
+     * Returns null if not an interesting method (i.e. from test class).
+     */
+    private String getMethodName(Map<Integer, String> compileIdMap, String line) {
+        Pattern pattern = Pattern.compile("compile_id='(\\d+)'");
+        Matcher matcher = pattern.matcher(line);
```
again, precompiled pattern can be saved into a (in this case) static field and reused.

iignatev commented on test/lib/jdk/test/lib/hotspot/ir_framework/IRMatcher.java:
```diff
@@ -0,0 +1,562 @@
+            String line = lines[i].trim();
+            String[] splitComma = line.split(",");
+            if (splitComma.length < 2) {
+                throw new TestFrameworkException("Invalid IR match rule encoding");
```
will it make sense to include the line-offender into the exception message here?

iignatev commented on test/lib/jdk/test/lib/hotspot/ir_framework/IRMatcher.java:
```diff
@@ -0,0 +1,562 @@
+                throw new TestFrameworkException("Invalid IR match rule encoding");
+            }
+            String testName = splitComma[0];
+            Integer[] irRulesIdx = new Integer[splitComma.length - 1];
```
you can actually use an array of int here, so there will be less wasted memory and no unboxing later on

iignatev commented on test/lib/jdk/test/lib/hotspot/ir_framework/IRMatcher.java:
```diff
@@ -0,0 +1,562 @@
+     */
+    private void parseHotspotPidFile() {
+        Map<Integer, String> compileIdMap = new HashMap<>();
+        try (BufferedReader br = new BufferedReader(new FileReader(new File(System.getProperty("user.dir") + File.separator + hotspotPidFileName)))) {
```
more idiomatic/modern version would be
```suggestion
try (BufferedReader br = Files.newBufferedReader(Paths.get(System.getProperty("user.dir"), hotspotPidFileName))) {
```

I actually not sure if you really need `$user.dir`, won't hotspot_pid file get generated in the current dir?

iignatev replied:
`Paths.get(hotspotPidFileName)` returns you a `Path` to `hotspotPidFileName` in cwd.

iignatev reviewed (changes requested):
next portion of my comments, stay tuned for more to come :)

-- Igor

PS I must say that by some reason github's 'file changed' tab is unbelievably slow for me in this particular PR, so it's somewhat painful to even just find the place I want to associate my comment with (not to speak about actually using it to browse/read the code)

chhagedorn commented on test/lib/jdk/test/lib/hotspot/ir_framework/CompLevel.java:
```diff
@@ -0,0 +1,134 @@
+     *      <li><p>{@link DontCompile @DontCompile}: Prevents any compilation of the associated helper method.</li>
+     *  </ul>
+     */
+    ANY(-2),
```
I'll change that as soon as it is integrate.

chhagedorn reviewed (commented):
Thanks a lot to @iignatev for your careful review and also to @vnkozlov! I will update this PR incrementally regarding the different review comments. I'll start with handling the comments directly in the code. I waited with some of the changes if there is still something open for discussion/follow-up.

chhagedorn commented:
> * although having javadoc for testlibraries is highly desirable, I don't think we should check in the generated HTML files
> * the same goes for `README.html` generated from `README.md`

Okay, I will remove them. Does it make sense to still have the HTML files somewhere in the web, for example, on my cr.openjdk? 

> * this library is hotspot-centric, I highly doubt that it will be used by any tests outside of the hotspot test base, hence the more appropriate location for it would be inside `test/hotspot/jtreg/testlibrary`.
> * I assume `test/lib/jdk/test/lib/hotspot/ir_framework/tests/` are the tests for the framework, in that case they should be in `test/lib-test`, if we stick to `test/lib` as the location for the library, or in `test/hotspot/jtreg/testlibrary_tests`, if we move it to `test/hotspot`

That makes sense to move everything to `test/hotspot/jtreg/testlibrary`. Right, the `test/lib/jdk/test/lib/hotspot/ir_framework/tests/` are only tests for the framework itself and should not be run as part of tier testing each time (does not make much sense) but only when the framework is actually modified. I guess that's still the case when putting them in `test/hotspot/jtreg/testlibrary_tests`?

I will do this things in a separate commit and adjust the README.md file accordingly (has links to the Javadoc files).

iignatev reviewed (commented):
@chhagedorn , I haven't finished reviewing the patch, but I figure it would be better to re-review the patch after you address the prev comments, so I will post the few pending comments I had and will return to review the whole patch after we finish discussing prev. comments.

Cheers,
-- Igor

iignatev commented:
> > * although having javadoc for testlibraries is highly desirable, I don't think we should check in the generated HTML files
> > * the same goes for `README.html` generated from `README.md`
> 
> Okay, I will remove them. Does it make sense to still have the HTML files somewhere in the web, for example, on my cr.openjdk?
> 
> > * this library is hotspot-centric, I highly doubt that it will be used by any tests outside of the hotspot test base, hence the more appropriate location for it would be inside `test/hotspot/jtreg/testlibrary`.
> > * I assume `test/lib/jdk/test/lib/hotspot/ir_framework/tests/` are the tests for the framework, in that case they should be in `test/lib-test`, if we stick to `test/lib` as the location for the library, or in `test/hotspot/jtreg/testlibrary_tests`, if we move it to `test/hotspot`
> 
> That makes sense to move everything to `test/hotspot/jtreg/testlibrary`. Right, the `test/lib/jdk/test/lib/hotspot/ir_framework/tests/` are only tests for the framework itself and should not be run as part of tier testing each time (does not make much sense) but only when the framework is actually modified. Is this still the case when putting them in `test/hotspot/jtreg/testlibrary_tests` (i.e. not executed unless run manually)?

`test/hotspot/jtreg/testlibrary_tests` are seldomly run as part of `:hotspot_misc` test group, yet I don't think it's an issue. the benefit of running testlibrary tests is to gain confidence that the tests which use these libraries are reliable. I'd also disagree that `ir_framework/tests` should be run *only* when the framework is actually, they should also be run when its dependencies are changed, and the framework highly depends on hotspot, so to be sure that the framework didn't get broken after changes in c2, whitebox, etc, you do need run these tests more often.

Thanks,
-- Igor

JohnTortugo commented on test/lib/jdk/test/lib/hotspot/ir_framework/examples/BaseTestExample.java:
```diff
@@ -0,0 +1,89 @@
+    int iFld;
+
+    public static void main(String[] args) {
+        TestFramework.run(); // equivalent to TestFramework.run(TestSimpleTest.class)
```
Did you mean same as `TestFramework.run(BaseTestExample.class)` ?

JohnTortugo commented on test/lib/jdk/test/lib/hotspot/ir_framework/examples/CheckedTestExample.java:
```diff
@@ -0,0 +1,148 @@
+public class CheckedTestExample {
+
+    public static void main(String[] args) {
+        TestFramework.run(); // equivalent to TestFramework.run(TestSimpleTest.class)
```
Did you mean same as `TestFramework.run(CheckedTestExample.class)` ?

JohnTortugo commented on test/lib/jdk/test/lib/hotspot/ir_framework/examples/CustomRunTestExample.java:
```diff
@@ -0,0 +1,167 @@
+public class CustomRunTestExample {
+
+    public static void main(String[] args) {
+        TestFramework.run(); // equivalent to TestFramework.run(TestSimpleTest.class)
```
Did you mean same as `TestFramework.run(CustomRunTestExample.class)` ?

JohnTortugo commented on test/lib/jdk/test/lib/hotspot/ir_framework/examples/IRExample.java:
```diff
@@ -0,0 +1,172 @@
+    @IR(failOn = IRNode.STORE)
+    @IR(failOn = {IRNode.STORE, IRNode.LOOP}) // LOOP regex not found but STORE regex, letting the rule fail
+    @IR(failOn = {IRNode.LOOP, IRNode.STORE}) // Order does not matter
+    @IR(failOn = {IRNode.STORE, IRNode.LOAD}) // LOOP and STORE regex, letting the rule fail
```
LOOP regex?

JohnTortugo commented on test/lib/jdk/test/lib/hotspot/ir_framework/IRMethod.java:
```diff
@@ -0,0 +1,115 @@
+ * Helper class to store information about a method that needs to be IR matched.
+ */
+class IRMethod {
+    final private Method method;
```
The order of these keywords seems to be inverted.

iignatev commented on test/lib/jdk/test/lib/hotspot/ir_framework/TestFramework.java:
```diff
@@ -0,0 +1,1226 @@
+        if (exitCode != 0) {
+            System.err.println("--- OUTPUT TestFramework flag VM ---");
+            System.err.println(flagVMOutput);
+            throw new RuntimeException("\nTestFramework flag VM exited with " + exitCode);
```
👍 . I noticed that you have leading `\n` in other exceptions, do you plan to remove it from there as well?

iignatev commented on test/lib/jdk/test/lib/hotspot/ir_framework/TestFramework.java:
```diff
@@ -0,0 +1,1226 @@
+        }
+    }
+
+    static void check(boolean test, String failureMessage) {
+        if (!test) {
+            fail(failureMessage);
+        }
+    }
+
+    static void fail(String failureMessage) {
+        throw new TestFrameworkException("Internal Test Framework exception - please file a bug:\n" + failureMessage);
+    }
+
+    static void fail(String failureMessage, Throwable e) {
+        throw new TestFrameworkException("Internal Test Framework exception - please file a bug:\n" + failureMessage, e);
+    }
+}
```
I understand that you want to use different exception types to distinguish different kinds of failures, I just don't see lots of benefits in have `::check` and `::fail` methods, they are just `if (p) throw new X(m)` or `throw new X(m)`.

if you want `TestFrameworkException` to always have `Internal Test Framework exception - please file a bug`, you can prepend it in `TestFrameworkException::<init>`

chhagedorn reviewed (commented):
Updated review comments from @JohnTortugo and @iignatev related to code.

chhagedorn commented:
> > > * although having javadoc for testlibraries is highly desirable, I don't think we should check in the generated HTML files
> > > * the same goes for `README.html` generated from `README.md`
> > 
> > 
> > Okay, I will remove them. Does it make sense to still have the HTML files somewhere in the web, for example, on my cr.openjdk?
> > > * this library is hotspot-centric, I highly doubt that it will be used by any tests outside of the hotspot test base, hence the more appropriate location for it would be inside `test/hotspot/jtreg/testlibrary`.
> > > * I assume `test/lib/jdk/test/lib/hotspot/ir_framework/tests/` are the tests for the framework, in that case they should be in `test/lib-test`, if we stick to `test/lib` as the location for the library, or in `test/hotspot/jtreg/testlibrary_tests`, if we move it to `test/hotspot`
> > 
> > 
> > That makes sense to move everything to `test/hotspot/jtreg/testlibrary`. Right, the `test/lib/jdk/test/lib/hotspot/ir_framework/tests/` are only tests for the framework itself and should not be run as part of tier testing each time (does not make much sense) but only when the framework is actually modified. Is this still the case when putting them in `test/hotspot/jtreg/testlibrary_tests` (i.e. not executed unless run manually)?
> 
> `test/hotspot/jtreg/testlibrary_tests` are seldomly run as part of `:hotspot_misc` test group, yet I don't think it's an issue. the benefit of running testlibrary tests is to gain confidence that the tests which use these libraries are reliable. I'd also disagree that `ir_framework/tests` should be run _only_ when the framework is actually, they should also be run when its dependencies are changed, and the framework highly depends on hotspot, so to be sure that the framework didn't get broken after changes in c2, whitebox, etc, you do need run these tests more often.

I see your point. I have not considered the dependencies which can frequently change even though the framework does not. I therefore agree with you that we should run them at least from time to time. I will move everything to  `test/hotspot/jtreg/testlibrary` in a next step.

chhagedorn commented:
There is no decision, yet, (and open for discussion) about the location and package name for the framework and the framework internal tests. After discussing it offline with @iignatev, we think there are basically two good options:
1. Leave the framework in `/test/lib` and put the framework internal tests into `/test/lib-test`:
   - Pros: Only need to specify `@library /test/lib` in a JTreg test; the framework internal tests are run in tier1 before any other tests are run which depend on the framework ensuring correctness.
   - Cons: Clarity: The framework is intended to be used for HotSpot tests only (thus not exactly the right location in `/test/lib`); the framework internal tests might be run too often as part of tier1 (trade-off ensuring correctness vs. execution time).
2. Move the framework to `/test/hotspot/jtreg/testlibrary`, put the framework internal tests into `/test/hotspot/jtreg/testlibrary_tests`, and use package name `hotspot.test.lib.ir_framework`:
   - Pros: Clarity: The framework is only used for HotSpot tests (mainly compiler tests but could also be used for other tests).
   - Cons: A JTreg test needs to additionally specify `/testlibrary/ir_framework` like `@library /testlibrary/ir_framework /test/lib`; the framework internal tests are run more seldomly as part of `:hotspot_misc` (trade-off, see cons of first option).
 
What do others think about that?

TobiHartmann commented on test/lib/jdk/test/lib/hotspot/ir_framework/README.md:
```diff
@@ -0,0 +1,137 @@
+- `-DVerifyIR=false`: Explicitly disable IR verification. This is useful, for example, if some scenarios use VM flags that let `@IR` annotation rules fail and the user does not want to provide separate IR rules or add flag preconditions to the already existing IR rules.
+- `-DTest=test1,test2`: Provide a list of `@Test` method names which should be executed.
+- `-DExclude=test3`: Provide a list of `@Test` method names which should be excluded from execution.
+- `-DScenarios=1,2`: Provide a list of scenario indexes to specify which scenarios that should be executed.
```
"that" should be removed

TobiHartmann commented on test/lib/jdk/test/lib/hotspot/ir_framework/README.md:
```diff
@@ -0,0 +1,137 @@
+- `-DVerifyVM=true`: The framework runs the test VM with additional verification flags (slows the execution down).
+- `-DExcluceRandom=true`: The framework randomly excludes some methods from compilation. IR verification is disabled completely with this flag.
+- `-DFlipC1C2=true`: The framework compiles all `@Test` annotated method with C1 if a C2 compilation would have been applied and vice versa. IR verification is disabled completely with this flag.
+- `-DShuffleTests=false`: Disables the shuffling of tests (a shuffling is always done by default).
```
Maybe explain that "shuffling" means execution in random order.

TobiHartmann commented on test/lib/jdk/test/lib/hotspot/ir_framework/README.md:
```diff
@@ -0,0 +1,137 @@
+## 5. Future Work
+This framework is based on the idea of the currently present IR test framework in [Valhalla](https://github.com/openjdk/valhalla/blob/e9c78ce4fcfd01361c35883e0d68f9ae5a80d079/test/hotspot/jtreg/runtime/valhalla/inlinetypes/InlineTypesTest.java). This IR framework was used with great success in Valhalla and thus served as a foundation for this new IR framework.
+ 
+ The new framework supports all the features that are present in the Valhalla IR framework with the idea to replace it at some point. The initial design and feature set was kept simple and straight forward and serves well for small to medium sized tests. There are a lot of possibilities to further enhance the framework and make it more powerful. This can be tackled in additional RFEs. A few ideas ...
```
Again, I would remove the reference to the Valhalla framework. Looking through the changes, I see 13 references to "Valhalla", I think these should all be removed.

chhagedorn replied:
I removed all references except the ones in the testing section about the converted tests. I think these references are important to keep for now.

TobiHartmann commented on test/lib/jdk/test/lib/hotspot/ir_framework/README.md:
```diff
@@ -0,0 +1,137 @@
+- Investigate more verification possibilities based on additional/other flags (e.g. _-XX:+PrintIdealGraph_, _-XX:+TraceNewVectors_ etc.) ([JDK-8265200](https://bugs.openjdk.java.net/browse/JDK-8265200)).
+- More interface methods for verification in checked and custom run tests.
+- More stress/debug framework flags.
+- Additional check possibilities in `@IR` annotations. 
```
I would remove the "Future Work" section and migrate it completely to RFEs. Otherwise it will quickly be outdated anyway.

TobiHartmann reviewed (approved):
Great work, Christian! I'm glad we finally have the capability to write regression tests with IR verification.

Hard to review this in detail but I've added some comments. In any case, conversion of Valhalla tests to the new framework has proven that this is good to go. We can still fix problems and add new features once this is in.

chhagedorn commented on test/lib/jdk/test/lib/hotspot/ir_framework/Check.java:
```diff
@@ -0,0 +1,113 @@
+ */
+@Retention(RetentionPolicy.RUNTIME)
+public @interface Check {
+
```
Cleaned that up together with some other formatting.

chhagedorn commented:
> Great work, Christian! I'm glad we finally have the capability to write regression tests with IR verification.
> 
> Hard to review this in detail but I've added some comments. In any case, conversion of Valhalla tests to the new framework has proven that this is good to go. We can still fix problems and add new features once this is in.

Thank you Tobias for your review! I pushed an update with your review comments. I agree that we can still follow up with fixes and new features once this went it.

iignatev commented on test/lib/jdk/test/lib/hotspot/ir_framework/TestFramework.java:
```diff
@@ -0,0 +1,1044 @@
+                                       #############################################################
+                                        - To only run the failed tests use -DTest, -DExclude,
+                                          and/or -DScenarios.
+                                        - To also get the standard output of the test VM run with\s
```
why do you need `\s` here?

chhagedorn replied:
I needed this because `jcheck` complained about trailing whitespaces.

iignatev replied:
well, why do you want to have a trailing space here? the text block you current have produces the following multeline sting, which has a trailing space at 4th line:
```  #############################################################
   - To only run the failed tests use -DTest, -DExclude,
     and/or -DScenarios.
   - To also get the standard output of the test VM run with 
     -DReportStdout=true or for even more fine-grained logging
     use -DVerbose=true.
  #############################################################```

chhagedorn replied:
Hm, it somehow made sense when I added it back there but looking at it again it no longer does. I'll remove it in the next iteration.

iignatev commented on test/lib/jdk/test/lib/hotspot/ir_framework/TestFramework.java:
```diff
@@ -0,0 +1,1044 @@
+    private static final boolean PREFER_COMMAND_LINE_FLAGS = Boolean.getBoolean("PreferCommandLineFlags");
+    private static final boolean EXCLUDE_RANDOM = Boolean.getBoolean("ExcludeRandom");
+    private static final boolean REPORT_STDOUT = Boolean.getBoolean("ReportStdout");
+    private final boolean VERIFY_VM = Boolean.getBoolean("VerifyVM") && Platform.isDebugBuild();
+    private boolean VERIFY_IR = Boolean.parseBoolean(System.getProperty("VerifyIR", "true"));
```
it seems both these guys were meant to be `static final`

chhagedorn replied:
`VERIFY_VM` should be yes. But `VERIFY_IR` is only initialized and later updated in `disableIRVerificationIfNotFeasible()`. However, I suggest to rename this field into something more meaningful like `irVerificationPossible`.

iignatev commented on test/lib/jdk/test/lib/hotspot/ir_framework/TestFramework.java:
```diff
@@ -0,0 +1,1044 @@
+    private static boolean toggleBool;
+
+    private final Class<?> testClass;
+    private Set<Class<?>> helperClasses = null;
+    private List<Scenario> scenarios = null;
+    private Set<Integer> scenarioIndices = null;
+    private List<String> flags = null;
```
`null` is a default value for reference types, so there is no need to set it explicitly.

iignatev commented:
> There is no decision, yet, (and open for discussion) about the location and package name for the framework and the framework internal tests. After discussing it offline with @iignatev, we think there are basically two good options:
> 
> 1. Leave the framework in `/test/lib` and put the framework internal tests into `/test/lib-test`:
>    
>    * Pros: Only need to specify `@library /test/lib` in a JTreg test; the framework internal tests are run in tier1 before any other tests are run which depend on the framework ensuring correctness.
>    * Cons: Clarity: The framework is intended to be used for HotSpot tests only (thus not exactly the right location in `/test/lib`); the framework internal tests might be run too often as part of tier1 (trade-off ensuring correctness vs. execution time).
> 2. Move the framework to `/test/hotspot/jtreg/testlibrary`, put the framework internal tests into `/test/hotspot/jtreg/testlibrary_tests`, and use package name `hotspot.test.lib.ir_framework`:
>    
>    * Pros: Clarity: The framework is only used for HotSpot tests (mainly compiler tests but could also be used for other tests).
>    * Cons: A JTreg test needs to additionally specify `/testlibrary/ir_framework` like `@library /testlibrary/ir_framework /test/lib`; the framework internal tests are run more seldomly as part of `:hotspot_misc` (trade-off, see cons of first option).
> 

there is also 3rd alternative, move the framework to `/test/hotspot/jtreg/compiler/` and use `compiler.ir_framework` or `compiler.lib.ir_framework` as the package name. that will make it even clearer that this is a compiler-specific library, and its users are going to be in `/test/hotspot/jtreg/compiler/`. I understand that there can be some runtime (or other) tests that might benefit from this library, but in the end, they won't be clear runtime tests, so `t/h/j/compiler` might be a better location for them anyways. the `@library` tag will still have to contain two paths, but these will be usual paths as in many compiler tests : `@library / /test/lib`.
the tests can be placed in `/test/hotspot/jtreg/testlibrary_tests` as in the 2nd option, with the same trade-off.

-- Igor

JohnTortugo commented on test/lib/jdk/test/lib/hotspot/ir_framework/TestFrameworkExecution.java:
```diff
@@ -0,0 +1,1452 @@
+            switch (random.nextInt() % 3) {
+                case 1 -> {
+                    level = CompLevel.C1;
+                    WHITE_BOX.makeMethodNotCompilable(ex, CompLevel.C2.getValue(), false);
```
I'm curious why two calls (with different values for the last parameter) are needed for these?

chhagedorn replied:
The parameter is for [osr](https://github.com/openjdk/jdk/blob/dc323a933401e4e540cdc416de5260923f42ba7f/src/hotspot/share/prims/whitebox.cpp#L928-L937): We need to exclude the method for osr and for normal compilations.

JohnTortugo commented on test/lib/jdk/test/lib/hotspot/ir_framework/TestFrameworkExecution.java:
```diff
@@ -0,0 +1,1452 @@
+        return compLevel;
+    }
+
+    private static CompLevel flipCompLevel(CompLevel compLevel) {
```
Wouldn't be better to put this on the CompLevel enum itself?

JohnTortugo commented on test/lib/jdk/test/lib/hotspot/ir_framework/Run.java:
```diff
@@ -0,0 +1,107 @@
+@Retention(RetentionPolicy.RUNTIME)
+public @interface Run {
+    /**
+     * The associated {@link Test @Test} methods (one or more) for for this {@code @Run} annotated run method.
```
Typo: "for for"

chhagedorn commented on test/lib/jdk/test/lib/hotspot/ir_framework/CompLevel.java:
```diff
@@ -0,0 +1,133 @@
+    /**
+     *  Compilation level 1: C1 compilation without any profile information.
+     */
+    C1(1),
```
I initially had `C1_SIMPLE` but then replaced it with `C1` as `@DontCompile` can only exclude a compilation for a compiler and not a specific level. However, while thinking about it, it might be better to not share `C1/C1_SIMPLE` but use a separate annotation for `@DontCompile`. I added a new `Compiler` annotation class and updated the usages and checks.

chhagedorn commented on test/lib/jdk/test/lib/hotspot/ir_framework/DontCompile.java:
```diff
@@ -0,0 +1,49 @@
+ * Prevent a compilation of the annotated <b>helper method</b> (not specifying {@link Test @Test},
+ * {@link Check @Check} or {@link Run @Run}):
+ *
+ * <ul>
+ *     <li><p>{@link CompLevel#ANY} (default): No C1 or C2 compilation.</li>
+ *     <li><p>{@link CompLevel#C1}: No C1 compilation, C2 compilation still possible.</li>
+ *     <li><p>{@link CompLevel#C2}: No C2 compilation, C1 compilation still possible.</li>
+ *     <li><p>The usage of any other compilation level is forbidden and results in a
+ *            {@link TestFormatException TestFormatException}.</li>
+ * </ul>
```
I was thinking about the same thing while processing an earlier review comment above :) So, it indeed makes sense to use a separate enum for it.

chhagedorn commented on test/lib/jdk/test/lib/hotspot/ir_framework/IRMatcher.java:
```diff
@@ -0,0 +1,488 @@
+ */
+class IRMatcher {
+    private static final boolean PRINT_IR_ENCODING = Boolean.parseBoolean(System.getProperty("PrintIREncoding", "false"));
+    private static final Pattern irEncodingPattern =
```
I updated this and other inconsistencies in the naming of `static final` variables.

chhagedorn commented on test/lib/jdk/test/lib/hotspot/ir_framework/Scenario.java:
```diff
@@ -0,0 +1,146 @@
+
+    static {
+        if (!SCENARIOS_PROPERTY.isEmpty()) {
+            System.out.println(Arrays.toString(SCENARIOS_PROPERTY.split("\\s*,\\s*")));
```
Good catch. It was a leftover.

chhagedorn commented on test/lib/jdk/test/lib/hotspot/ir_framework/TestFrameworkExecution.java:
```diff
@@ -0,0 +1,1452 @@
+    }
+
+    private void processControlAnnotations(Class<?> clazz) {
+        if (!XCOMP) {
```
This handling with `XCOMP` (and other related ones) seem to be wrong and only worked like that in the old Valhalla framework. I don't think it is correct like that anymore. I discussed it with @TobiHartmann and thus added a new debug flag `-DIgnoreCompilerControls` to achieve the same functionality as in the old framework. I will push this update in a separate commit.

chhagedorn commented on test/lib/jdk/test/lib/hotspot/ir_framework/TestFrameworkExecution.java:
```diff
@@ -0,0 +1,1452 @@
+        Random random = Utils.getRandomInstance();
+        boolean exclude = random.nextBoolean();
+        CompLevel level = CompLevel.ANY;
+        if (exclude) {
```
I refactored the code relating to `ExcludeRandom`.

iignatev commented on test/lib/jdk/test/lib/hotspot/ir_framework/TestFramework.java:
```diff
@@ -0,0 +1,1044 @@
+    private final ProcessBuilder process;
+    private final String hotspotPidFileName;
+
+    JVMOutput(OutputAnalyzer oa, Scenario scenario, ProcessBuilder process) {
```
instead of passing a ProcessBuilder, you can either pass a command-line or better just past prepared flags, and make it JVMOutput's (you will need a better name) responsibility to execute the process, create OutputAnalyzer, etc. 

```suggestion
    JVMOutput(List<String> cmds, Scenario scenario) {
    var pb = ProcessTools.createJavaProcessBuilder(cmds);
    try {
            // Calls 'main' of TestFrameworkExecution to run all specified tests with commands 'cmds'.
            // Use executeProcess instead of executeTestJvm as we have already added the JTreg VM and
            // Java options in prepareTestVMFlags().
            this.oa = ProcessTools.executeProcess(process);
        } catch (Exception e) {
            throw new TestFrameworkException("Error while executing Test VM", e);
        }
        this.cmd = pb.command();
        ...
```

and then runTest will be smth like
```
    private void runTestVM(List<String> additionalFlags) {
          List<String> cmds = prepareTestVMFlags(additionalFlags);
          socket.start();
          JVMOutput output = new JVMOutput(cmds, scenario);
          ...
```

chhagedorn replied:
I did a refactoring and introduced new classes `FlagVMProcess` and `TestVMProcess`. I also did a renaming of `TestFrameworkPrepareFlags` -> `FlagVM` and `TestFrameworkExecution` -> `TestVM` to make it more clear whats going on.

chhagedorn replied:
I did a refactoring and introduced new classes `FlagVMProcess` and `TestVMProcess`. I also did a renaming of `TestFrameworkPrepareFlags` -> `FlagVM` and `TestFrameworkExecution` -> `TestVM` to make it more clear whats going on.

chhagedorn commented on test/lib/jdk/test/lib/hotspot/ir_framework/TestFrameworkPrepareFlags.java:
```diff
@@ -0,0 +1,120 @@
+        String encoding = TestFramework.TEST_VM_FLAGS_START + System.lineSeparator()
+                          + String.join(TestFramework.TEST_VM_FLAGS_DELIMITER, flags)
+                          + System.lineSeparator() + TestFramework.TEST_VM_FLAGS_END;
+        TestFrameworkSocket.write(encoding, "flag encoding");
```
Okay, yes it would make it easier. But the socket can be kept for the communication between the test VM and the driver VM?

iignatev replied:
I don't have a strong opinion here. let's start w/ what you have now, we can always change it later if we find that its complexity doesn't bring much value.

chhagedorn replied:
Sounds good. I'll push an update with the changes tomorrow. Thanks for your review!

chhagedorn replied:
I moved to a file-based approach for the flag VM and cleaned the socket up for the sole usage for the test VM.

chhagedorn commented on test/lib/jdk/test/lib/hotspot/ir_framework/TestInfo.java:
```diff
@@ -0,0 +1,94 @@
+     * Return a boolean indicating if the framework skipped a compilation after the warm-up due to VM flags not
+     * allowing a compilation on the requested level in {@link Test#compLevel()}.
+     *
+     * @return {@code true} if the framework compiled the test;
```
Yes, they are implemented slightly differently.

chhagedorn commented:
I did a slightly bigger refactoring reflecting the review comments by @iignatev:
- Extracting classes into own Java files
- Class renamings
- Moving to file-based communication for the flag VM to simplify the socket usage for the test VM (required some clean-ups)
- Removed some `TestFramework` public methods as discussed earlier in this PR.

Missing (I will tackle this tomorrow):
- separating classes into own (sub)packages
- trying to move tests to `/test/hotspot/jtreg/testlibrary_tests` and the framework to `/test/hotspot/jtreg/compiler` as suggested in an earlier comment

chhagedorn commented on test/lib/jdk/test/lib/hotspot/ir_framework/TestVMException.java:
```diff
@@ -0,0 +1,45 @@
+
+    TestVMException(String exceptionInfo) {
+        super("There were one or multiple errors. Please check stderr for more information.");
+        this.exceptionInfo = exceptionInfo;
```
There were errors in some cases when the message was too long. Thus, I used a separate field for it which I then print to the stderr.

chhagedorn commented:
> I've reviewed all the files but `examples` and `tests`, for them I'd actually recommend you to put `examples` into `tests` (possible in a subdir), and just update readme to say that the examples can be found in test directory.

Thanks a lot Igor! I'll consider to move the `examples` as well tomorrow.

chhagedorn commented:
As mentioned above, I moved the framework classes and tests and updated the packages. I also split the classes into subpackages to structure the code better.  A summary of the package structure can be found in the updated README file in section 5.

I put this PR on hold as I'm away until the end of the month.

TobiHartmann commented on test/hotspot/jtreg/compiler/lib/ir_framework/Compiler.java:
```diff
@@ -0,0 +1,63 @@
+    /**
+     * Selecting both the C1 and C2 compiler. This must be in sync with hotspot/share/compiler/compilerDefinitions.hpp.
+     */
+    ANY(-2),
```
After the removal of AOT, this should be `-1`.

TobiHartmann commented on test/hotspot/jtreg/compiler/lib/ir_framework/CompLevel.java:
```diff
@@ -0,0 +1,171 @@
+     *      <li><p>{@link DontCompile @DontCompile}: Prevents any compilation of the associated helper method.</li>
+     *  </ul>
+     */
+    ANY(-2),
```
After the removal of AOT, this should be `-1`.

chhagedorn commented:
I'm back to work and "reopen" this PR again.

@TobiHartmann Good point! I merged the master branch and fixed this to `-1` together with some other test failures that emerged after the update.

chhagedorn commented:
/contributor remove @TobiHartmann

chhagedorn commented:
/contributor add @chhagedorn

chhagedorn commented:
/contributor add @TobiHartmann

iignatev commented on test/hotspot/jtreg/compiler/lib/ir_framework/README.md:
```diff
@@ -0,0 +1,135 @@
+There are various ways how to set up and run a test within the `main()` method of a JTreg test. These are described and can be found in the [TestFramework](./TestFramework.java) class.
+   
+## 2. Features
+ The framework offers various annotations and flags to control how your test code should be invoked and being checked. This section gives an overview over all these features.
```
redundant space at the beginning of the line

iignatev commented on test/hotspot/jtreg/compiler/lib/ir_framework/README.md:
```diff
@@ -0,0 +1,135 @@
+- `compiler.lib.ir_framework.shared`: These classes can be called from either the driver, flag, or test VM.
+
+## 6. Summary
+ The initial design and feature set was kept simple and straight forward and serves well for small to medium sized tests. There are a lot of possibilities to further enhance the framework and make it more powerful. This can be tackled in additional RFEs. A few ideas can be found as subtasks of the [initial RFE](https://bugs.openjdk.java.net/browse/JDK-8254129) for this framework.
```
redundant space:
```suggestion
The initial design and feature set was kept simple and straight forward and serves well for small to medium sized tests. There are a lot of possibilities to further enhance the framework and make it more powerful. This can be tackled in additional RFEs. A few ideas can be found as subtasks of the [initial RFE](https://bugs.openjdk.java.net/browse/JDK-8254129) for this framework.
```

iignatev reviewed (approved):
LGTM, modulo two nits in README file and the location of tests and examples: I'd prefer to place them into `/test/hotspot/jtreg/testlibrary_tests/ir_framework` and add explicit reference to that location in README file.

chhagedorn commented:
Thanks a lot Igor for your review(s)! I pushed an update accordingly together with some last test fixes. I'm running some more testing over the weekend, also together with the converted and updated Valhalla tests. If everything looks fine, I will integrate it on Monday.

chhagedorn commented:
Testing looked good! I'm doing some last checks before integrating it later today.

Thanks to all for carefully reviewing it and providing valuable feedback.

chhagedorn commented:
/integrate
